include/highfive/h5easy_bits/H5Easy_Eigen.hpp

```diff
@@ -58,66 +84,68 @@ struct io_impl<
+                               const DumpOptions& options) {
+        using row_major_type = typename types<T>::row_major;
         using value_type = typename std::decay<T>::type::Scalar;
-        detail::createGroupsToDataSet(file, path);
-        DataSet dataset = file.createDataSet<value_type>(path, DataSpace(shape(data)));
-        write(dataset, data);
-        file.flush();
-        return dataset;
-    }
-
-    static DataSet overwrite(File& file, const std::string& path, const T& data) {
-        DataSet dataset = file.getDataSet(path);
-        if (dataset.getDimensions() != shape(data)) {
-            throw detail::error(file, path, "H5Easy::dump: Inconsistent dimensions");
+        row_major_type row_major(data);
+        DataSet dataset = initDataset<value_type>(file, path, shape(data), options);
+        dataset.write_raw(row_major.data());
+        if (options.flush()) {
+            file.flush();
         }
-        write(dataset, data);
-        file.flush();
         return dataset;
     }
 
-    static T load(const File& file, const std::string& path) {
+    inline static T load(const File& file, const std::string& path) {
         DataSet dataset = file.getDataSet(path);
         std::vector<typename T::Index> dims = shape(file, path, dataset, T::RowsAtCompileTime);
         T data(dims[0], dims[1]);
         dataset.read(data.data());
-
         if (data.IsVectorAtCompileTime || data.IsRowMajor) {
             return data;
         }
+        using col_major = typename types<T>::col_major;
+        return col_major(data.data(), dims[0], dims[1]);
+    }
 
-        // convert to ColMajor if needed (HDF5 always stores row-major)
-        return Eigen::Map<Eigen::Array<
-            typename T::Scalar,
-            T::RowsAtCompileTime,
-            T::ColsAtCompileTime,
-            T::ColsAtCompileTime == 1 ? Eigen::ColMajor : Eigen::RowMajor,
-            T::MaxRowsAtCompileTime,
-            T::MaxColsAtCompileTime>>(data.data(), dims[0], dims[1]);
+    inline static Attribute dumpAttribute(File& file,
+                                          const std::string& path,
+                                          const std::string& key,
+                                          const T& data,
+                                          const DumpOptions& options) {
+        using row_major_type = typename types<T>::row_major;
+        using value_type = typename std::decay<T>::type::Scalar;
+        row_major_type row_major(data);
+        Attribute attribute = initAttribute<value_type>(file, path, key, shape(data), options);
+        attribute.write_raw(row_major.data());
+        if (options.flush()) {
+            file.flush();
+        }
+        return attribute;
+    }
+
+    inline static T loadAttribute(const File& file,
+                                  const std::string& path,
+                                  const std::string& key) {
+        DataSet dataset = file.getDataSet(path);
```

Comment:
Some of these methods may return a ref. To avoid copies and potentially improve performance, consider declaring const-ref when applicable.

Reply:
Do you mean using `const DataSet dataset = file.getDataSet(path);` etc. ? Here I trust your expertise. But let's merge this first and do a separate PR. This make the extra API features public. Also, once this PR is merged I can finish https://github.com/BlueBrain/HighFive/pull/343 . Thereafter there are only internal changes.

A related point is that a minor release could be considered soon, as there are quite some new features (also in https://github.com/BlueBrain/HighFive/pull/346 ) and it would be great to make them available to (conda) package users.

Reply:
None of the actual data is actually copied, though? I think in that case the penalty here is negligible, even though I agree that `const Dataset&` would be more appropriate.

Reply:
Let's do it in a separate PR!